File: Data Structures & Algorithms/search-2d-matrix/submission-0.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])
        left = 0
        right = rows * cols - 1

        while left <= right:
            mid = left + ((right - left) // 2)

            if target > matrix[mid // cols][mid % cols]:
                left = mid + 1
            elif target < matrix[mid // cols][mid % cols]:
                right = mid - 1
            else:
                return True
        
        return False

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        // pivot
        int minIndex = findMin(nums);

        // check if the target is to the left of pivor or right of pivot
        int left = 0;
        int right = nums.length - 1;

        if(target >= nums[minIndex] && target <= nums[right]) {
            left = minIndex;
        } else {
            right = minIndex - 1;
        }

        // traditional binary search
        while(left <= right) {
            int mid = (left + right) >>> 1;
            if(nums[mid] == target) {
                return mid;
            } else if(nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return -1;
    }
    public int findMin(int[] nums) {
        int left = 0;
        int right = nums.length - 1;

        while(left < right) {
            int mid = (left + right) >>> 1;

            if(nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}
